sycl/doc/extensions/LevelZeroBackend/LevelZeroBackend.md

```diff
@@ -0,0 +1,119 @@
+for a specific SYCL object. It is currently supported for SYCL ```platform, device, context, queue, event``` and ```program``` classes.
+There is also a free-function defined in ```cl::sycl``` namespace that can be used instead of the member function:
+``` C++
+template <backend BackendName, class SyclObjectT>
+auto get_native(const SyclObjectT &Obj) ->
+	typename interop<BackendName, SyclObjectT>::type;
+```
+### 4.3 Construct a SYCL object from a Level-Zero handle
+        
+The following free functions defined in the ```cl::sycl::level_zero``` namespace allow an application to create
+a SYCL object that encapsulates a corresponding Level-Zero object:
+
+| Level-Zero interoperability function |Description|
+|-------------|:------------|
+|``` make<platform>(ze_driver_handle_t);```|Constructs a SYCL platform instance from a Level-Zero ```ze_driver_handle_t```.|
+|``` make<device>(const platform &, ze_device_handle_t);```|Constructs a SYCL device instance from a Level-Zero ```ze_device_handle_t```. The platform argument gives a SYCL platform, encapsulating a Level-Zero driver supporting the passed Level-Zero device.|
+|``` make<context>(const vector_class<device> &, ze_context_handle_t);```| Constructs a SYCL context instance from a Level-Zero ```ze_context_handle_t```. The context is created against the devices passed in. There must be at least one device given and all the devices must be from the same SYCL platform and thus from the same Level-Zero driver.|
+|``` make<queue>(const context &, ze_command_queue_handle_t);```| Constructs a SYCL queue instance from a Level-Zero ```ze_command_queue_handle_t```. The context argument must be a valid SYCL context encapsulating a Level-Zero context. The queue is attached to the first device in the passed SYCL context.|
+|``` make<program>(const context &, ze_module_handle_t);```| Constructs a SYCL program instance from a Level-Zero ```ze_module_handle_t```. The context argument must be a valid SYCL context encapsulating a Level-Zero context. The Level-Zero module must be fully linked (i.e. not require further linking through [```zeModuleDynamicLink```](https://spec.oneapi.com/level-zero/latest/core/api.html?highlight=zemoduledynamiclink#_CPPv419zeModuleDynamicLink8uint32_tP18ze_module_handle_tP28ze_module_build_log_handle_t)),and thus the SYCL program is created in the "linked" state.|
+
+NOTE: We shall consider adding other interoperability as needed, if possible.
+                
+### 4.4 Level-Zero handles' ownership and lifetime
+        
+The Level-Zero runtime doesn't do reference-counting of its objects, so it is crucial to adhere to these
+practices of how Level-Zero handles are manged.
+                
+#### 4.4.1 SYCL runtime takes ownership
+                
+Whenever the application creates a SYCL object from the corresponding Level-Zero handle via one of the ```make<T>()``` functions,

```

Comment:
Should possible thread safety issues be mentioned? It seems unsafe to use L0 API which can modify a L0 handle while SYCL RT owns it as there is no guarantee that SYCL RT doesn't make any L0 call using the same handle at the same time.

Reply:
I thought the L0 APIs were all thread-safe.  I didn't check them all, but the documentation has this bullet point for all the ones I did check:

> The application may call this function from simultaneous threads.

Are there L0 APIs that are not thread-safe?

Reply:
https://spec.oneapi.com/level-zero/latest/core/INTRO.html#multithreading-and-concurrency

Reply:
Hmm, that seems like it will make it very difficult for the application to use any Level Zero API in a SYCL application.  I presume the DPC++ implementation has internal threads which might call Level Zero at any time during the application's execution, right?  In this case, there's not really any way for the application to know when it can call Level Zero without colliding with one of those internal calls.

At first, I thought we could say something like this:

> The application must take care when calling any Level-Zero API because not all Level-Zero APIs are thread safe
> and the SYCL runtime may itself call Level-Zero at any point in the application's execution.  Therefore, the
> application may only safely call Level-Zero APIs that are documented as safe to be called simultaneously from
> multiple threads.

However, even that is not strictly safe because the Level Zero documentation says this:

> APIs are not thread-safe with other APIs that use the same driver’s object handle
>
> * the application must ensure multiple threads do not enter these APIs when the handle is the same

Consider a Level Zero API like `zeDeviceGetProperties()`, which is documented as "The application may call this function from simultaneous threads".  You might think it is safe to call this function without any locking, but the statement I quote above means this is not strictly true.  What if the DPC++ runtime happens to call, for example, `zeDeviceGetStatus()` with the same device handle at the same time?  According to the Level Zero spec, this is not safe because `zeDeviceGetProperties()` and `zeDeviceGetStatus()` are different APIs, so it it not safe to call them with the same device handle from different threads.

My guess is that the Level Zero driver is actually safe to call in cases like this, but the documentation does not guarantee this safety.  One solution, perhaps, is to reconsider the Level Zero documentation and rewrite it to provide stronger thread safety.

Other possible solutions would require changes to DPC++:

* We could expose a new DPC++ API that acquires / releases a global lock which protects any of the Level Zero APIs.  Applications would be responsible for acquiring this lock before making any Level Zero call.
* We could expose an API that acquires / releases separate locks for each Level Zero handle.

Reply:
> I presume the DPC++ implementation has internal threads which might call Level Zero at any time during the application's execution, right?

Yes, it can happen.

Reply:
> I thought the L0 APIs were all thread-safe.

The Level-Zero API is **not thread-safe**, but it *is free-threaded* meaning that the access to any global/shared state is guarded while it is is expected of applications that the input arguments aren't used simultaneously from different threads. That synchronization/locking is done in the SYCL RT and Level-Zero plugin.

> very difficult for the application to use any Level Zero API in a SYCL application.

Practically speaking, and taking into account that SYCL RT **takes ownership** of the Level-Zero handles, the app should not attempt further direct use of those handles, except in API that is known to get immutable data, like device info, probably.

Thanks @romanovvlad for pointing out the thread-safety aspect. I do agree we need to have a chapter about that and will add one now.